mypy/stubtest.py

```diff
@@ -369,25 +358,29 @@ def verify_mypyfile(
         if not o.module_hidden and (not is_probably_private(m) or hasattr(runtime, m))
     }
 
-    imported_symbols = _get_imported_symbol_names(runtime)
-
     def _belongs_to_runtime(r: types.ModuleType, attr: str) -> bool:
         """Heuristics to determine whether a name originates from another module."""
         obj = getattr(r, attr)
         if isinstance(obj, types.ModuleType):
             return False
-        if callable(obj):
-            # It's highly likely to be a class or a function if it's callable,
-            # so the __module__ attribute will give a good indication of which module it comes from
-            try:
-                obj_mod = obj.__module__
-            except Exception:
-                pass
-            else:
-                if isinstance(obj_mod, str):
-                    return bool(obj_mod == r.__name__)
-        if imported_symbols is not None:
-            return attr not in imported_symbols
+
+        symbols = _module_symbol_table(r)
+        if symbols is not None:
+            if attr in {sym.get_name() for sym in symbols.get_symbols() if sym.is_imported()}:
+                # symtable says we got this from another module
+                return False
+
+            # but we can't just return True here, because symtable doesn't know about star imports
+            if attr in {sym.get_name() for sym in symbols.get_symbols() if sym.is_assigned()}:
+                # symtable knows we assigned this symbol in the module
+                return True
+        try:
+            obj_mod = obj.__module__
+        except Exception:
+            pass
+        else:
+            if isinstance(obj_mod, str):
+                return bool(obj_mod == r.__name__)
```

Comment:
I'm still not wild about relying on `__module__` like this. For pure-Python classes, it seems to me like it's more often incorrect than correct for anything except class objects and functions. Most instances just inherit the `__module__` attribute from their class, so it doesn't tell you anything about where the instance was created, just where the class of that instance was created:

```pycon
Python 3.12.1 (tags/v3.12.1:2305ca5, Dec  7 2023, 22:03:25) [MSC v.1937 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import logging, re, ipaddress, collections, threading, calendar
>>> logging.Logger("foo").__module__
'logging'
>>> re.compile("foo").__module__
're'
>>> ipaddress.IPv4Address(0).__module__
'ipaddress'
>>> collections.ChainMap().__module__
'collections'
>>> threading.Thread().__module__
'threading'
>>> calendar.Calendar().__module__
'calendar'
```

I guess I'm okay with this if we're sure it gets the number of false positives down, but I think we should at least add a comment saying that it's an unreliable heuristic